### ISSUE TITLE
add IDTOKENS authentication to TW

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6903]
make it possible for TW to submit to sched's using HTCondor tokens.
 @mmascher and @saqibhaleem please correct what I wrote in case I got it wrong

Need to change just one line
```
grep IDTOKENS v3.211008/slc7_amd64_gcc630/cms/crabtaskworker/v3.211008/lib/python2.7/site-packages/HTCondorUtils.py
htcondor.param['SEC_CLIENT_AUTHENTICATION_METHODS'] = 'FS,IDTOKENS,GSI' 
```

that will automatically pick a token from  `/home/crab3/.condor/tokens.d/` and use it.
We currently have one for ITB pool available and will use it for tests.
```
[crab3@crab-preprod-tw02 TaskManager]$ condor_token_list
Header: {"alg":"HS256","kid":"itb_pool_key"} Payload: {"iat":1638885389,"iss":"vocms0803.cern.ch:9618","jti":"f544c417ed884eba51ad29f6ef825433","sub":"crabtw@cms"} File: /home/crab3/.condor/tokens.d/itb_pool.token
```

One for production can be obtained from Saqib at the proper time.

**Comment by belforte** [comment]
Time to try this on my own.
I have done the changes on crab-dev-tw01 and copied there the above token.
But when TW tries to connect to ITB fails with this message in vocms068:/var/log/condor/SchedLog 
```
01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:36661> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS
01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:39907> did not result in a valid mapped user name, which is required for this command (478 ACT_ON_JOBS), so aborting.
```
137.138.159.63 is the IP for crab-dev-tw01

@saqib @mmascher maybe the token which you gave me is tied to a limited, specific, list of node names and crab-dev-tw01 is not there ?

**Comment by belforte** [comment]
indeed same thing works from crab-preprod-tw02. But we now use preprod instance for the py2/py3 migration and I would like to use our test(dev) instance for developing the IDTOKENS

**Comment by belforte** [comment]
hmmm.. I get same failure from crab-preprod-tw02, where we tested this together in December.
Any idea ?
SchedLog on vocms068:
```
01/14/22 18:39:28 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.52.227:42199> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
01/14/22 18:39:28 (pid:124668) DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS|AUTHENTICATE:1004:Failed to authenticate using FS|FS:1004:Unable to lstat(/tmp/FS_XXXUcuWOT)
01/14/22 18:39:28 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.52.227:44164> did not result in a valid mapped user name, which is required for this command (478 ACT_ON_JOBS), so aborting.
```
and 
[root@vocms068 condor]# host 137.138.52.227
227.52.138.137.in-addr.arpa domain name pointer crab-preprod-tw02.cern.ch.

**Comment by belforte** [comment]
anyhow. I am returning preprod instance to py2/3 testing. We need to make this work on crab-dev-tw01 as a start.

**Comment by belforte** [comment]
condor version in TW
```
[crab3@crab-dev-tw01 TaskManager]$ python3
Python 3.8.2 (default, Jan 22 2021, 17:57:37) 
[GCC 6.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import htcondor
>>> htcondor.version()
'$CondorVersion: 8.9.7 May 20 2020 BuildID: UW_Python_Wheel_Build $'
>>> 
[crab3@crab-dev-tw01 TaskManager]$ 
```

SchedLog with SCHED_DEBUG=D_SECURITY in 99_local_tweaks,
seems to complain about  "Error from client." but I have no logs client side only the dreaded exception in dagman fork
```
01/14/22 21:44:36 DC_AUTHENTICATE: received DC_AUTHENTICATE from <137.138.159.63:35317>
01/14/22 21:44:36 Inserting pre-auth metadata for TOKEN.
01/14/22 21:44:36 Inserting pre-auth metadata for TOKEN.
01/14/22 21:44:36 DC_AUTHENTICATE: generating BLOWFISH key for session vocms068:3205141:1642193076:30...
01/14/22 21:44:36 SECMAN: new session, doing initial authentication.
01/14/22 21:44:36 Returning to DC while we wait for socket to authenticate.
01/14/22 21:44:36 AUTHENTICATE: setting timeout for (unknown) to 20.
01/14/22 21:44:36 HANDSHAKE: in handshake(my_methods = 'TOKEN')
01/14/22 21:44:36 HANDSHAKE: handshake() - i am the server
01/14/22 21:44:36 HANDSHAKE: client sent (methods == 2048)
01/14/22 21:44:36 HANDSHAKE: i picked (method == 2048)
01/14/22 21:44:36 HANDSHAKE: client received (method == 2048)
01/14/22 21:44:36 PW.
01/14/22 21:44:36 Will return to DC because authentication is incomplete.
01/14/22 21:44:36 PASSWORD: entered authenticate_continue, state==100
01/14/22 21:44:36 PW: Server receiving 1.
01/14/22 21:44:36 Received: -1, 0(), 0
01/14/22 21:44:36 PW: Server received ERROR from client, propagating
01/14/22 21:44:36 PW: Server sending.
01/14/22 21:44:36 In server_send: -1.
01/14/22 21:44:36 Server send '', '', 0 0 0
01/14/22 21:44:36 PASSWORD: leaving authenticate_continue, state==101, return=2
01/14/22 21:44:36 AUTHENTICATE: auth would still block
01/14/22 21:44:36 Will return to DC to continue authentication..
01/14/22 21:44:36 PASSWORD: entered authenticate_continue, state==101
01/14/22 21:44:36 PW: Server receiving 2.
01/14/22 21:44:36 Error from client.
01/14/22 21:44:36 PASSWORD: leaving authenticate_continue, state==101, return=0
01/14/22 21:44:36 AUTHENTICATE: method 2048 (IDTOKENS) failed.
01/14/22 21:44:36 HANDSHAKE: in handshake(my_methods = 'TOKEN')
01/14/22 21:44:36 HANDSHAKE: handshake() - i am the server
01/14/22 21:44:36 HANDSHAKE: client sent (methods == 0)
01/14/22 21:44:36 HANDSHAKE: i picked (method == 0)
01/14/22 21:44:36 HANDSHAKE: client received (method == 0)
01/14/22 21:44:36 DC_AUTHENTICATE: authentication of <137.138.159.63:35317> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
01/14/22 21:44:36 DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS
```

**Comment by belforte** [comment]
I wonder if we should add 
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/condorpy3.spec
to the TW build
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/crabtaskworker.spec
like we had `condor` in the py2 build
Let's add @mapellidario to the thread

**Comment by saqibhaleem** [comment]
> Time to try this on my own. I have done the changes on crab-dev-tw01 and copied there the above token. But when TW tries to connect to ITB fails with this message in vocms068:/var/log/condor/SchedLog
> 
> ```
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:36661> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:39907> did not result in a valid mapped user name, which is required for this command (478 ACT_ON_JOBS), so aborting.
> ```
> 
> 137.138.159.63 is the IP for crab-dev-tw01
> 
> @saqibhaleem @mmascher maybe the token which you gave me is tied to a limited, specific, list of node names and crab-dev-tw01 is not there ?


> Time to try this on my own. I have done the changes on crab-dev-tw01 and copied there the above token. But when TW tries to connect to ITB fails with this message in vocms068:/var/log/condor/SchedLog
> 
> ```
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:36661> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS
> 01/14/22 17:57:36 (pid:124668) DC_AUTHENTICATE: authentication of <137.138.159.63:39907> did not result in a valid mapped user name, which is required for this command (478 ACT_ON_JOBS), so aborting.
> ```
> 
> 137.138.159.63 is the IP for crab-dev-tw01
> 
> @saqibhaleem @mmascher maybe the token which you gave me is tied to a limited, specific, list of node names and crab-dev-tw01 is not there ?

@belforte  We need to authorize token identity i.e. "crabtw" for schedd write operation.   I just added this line on vocms068.  

SCHEDD.ALLOW_WRITE= $(SCHEDD.ALLOW_WRITE),crabtw@cms

**Comment by belforte** [comment]
thanks @saqibhaleem but the error is still there. Also.. we tested this in December and it was working, was that line in condor config about ALLOW_WRITE present there and then removed ?

Reading log more carefully, I think there is no info all as to why TOKEN failed, and then simply server falls back to try PASSWORD which is where the "client error" is reported.

```
01/15/22 10:25:48 DC_AUTHENTICATE: received DC_AUTHENTICATE from <137.138.159.63:38769>
01/15/22 10:25:48 Inserting pre-auth metadata for TOKEN.
01/15/22 10:25:48 Inserting pre-auth metadata for TOKEN.
01/15/22 10:25:48 DC_AUTHENTICATE: generating BLOWFISH key for session vocms068:3205141:1642238748:8506...
01/15/22 10:25:48 SECMAN: new session, doing initial authentication.
01/15/22 10:25:48 Returning to DC while we wait for socket to authenticate.
01/15/22 10:25:48 AUTHENTICATE: setting timeout for (unknown) to 20.
01/15/22 10:25:48 HANDSHAKE: in handshake(my_methods = 'TOKEN')
01/15/22 10:25:48 HANDSHAKE: handshake() - i am the server
01/15/22 10:25:48 HANDSHAKE: client sent (methods == 2048)
01/15/22 10:25:48 HANDSHAKE: i picked (method == 2048)
01/15/22 10:25:48 HANDSHAKE: client received (method == 2048)
01/15/22 10:25:48 PW.
01/15/22 10:25:48 Will return to DC because authentication is incomplete.
01/15/22 10:25:48 PASSWORD: entered authenticate_continue, state==100
01/15/22 10:25:48 PW: Server receiving 1.
01/15/22 10:25:48 Received: -1, 0(), 0
01/15/22 10:25:48 PW: Server received ERROR from client, propagating
01/15/22 10:25:48 PW: Server sending.
01/15/22 10:25:48 In server_send: -1.
01/15/22 10:25:48 Server send '', '', 0 0 0
01/15/22 10:25:48 PASSWORD: leaving authenticate_continue, state==101, return=2
01/15/22 10:25:48 AUTHENTICATE: auth would still block
01/15/22 10:25:48 Will return to DC to continue authentication..
01/15/22 10:25:48 PASSWORD: entered authenticate_continue, state==101
01/15/22 10:25:48 PW: Server receiving 2.
01/15/22 10:25:48 Error from client.
01/15/22 10:25:48 PASSWORD: leaving authenticate_continue, state==101, return=0
01/15/22 10:25:48 AUTHENTICATE: method 2048 (IDTOKENS) failed.
01/15/22 10:25:48 HANDSHAKE: in handshake(my_methods = 'TOKEN')
01/15/22 10:25:48 HANDSHAKE: handshake() - i am the server
01/15/22 10:25:48 HANDSHAKE: client sent (methods == 0)
01/15/22 10:25:48 HANDSHAKE: i picked (method == 0)
01/15/22 10:25:48 HANDSHAKE: client received (method == 0)
01/15/22 10:25:48 DC_AUTHENTICATE: authentication of <137.138.159.63:38769> did not result in a valid mapped user name, which is required for this command (1112 QMGMT_WRITE_CMD), so aborting.
01/15/22 10:25:48 DC_AUTHENTICATE: reason for authentication failure: AUTHENTICATE:1003:Failed to authenticate with any method|AUTHENTICATE:1004:Failed to authenticate using IDTOKENS
```

do we have example on how to test this token w/o the burden of crab TW around it, to make sure that things are setup correctly on both sides to begin with ?

Could it be because in December we were using a python2 build for the test while this is python3 ?

**Comment by belforte** [comment]
I tried the "simple" `yum install condor` inside the TW container, but it does not bring condor_tokens* commands and  it is a 8.3.8 version. the condor which he have in old python2 TW (from COMP rpm's) is 9.2.0
the htcondor API that we have via COMP py3-htcondor is
```
[crab3@crab-dev-tw01 TaskManager]$ python3
Python 3.8.2 (default, Jan 22 2021, 17:57:37) 
[GCC 6.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import htcondor
>>> htcondor.version()
'$CondorVersion: 8.9.7 May 20 2020 BuildID: UW_Python_Wheel_Build $'
>>> 
```
could that be a problem ?

**Comment by belforte** [comment]
note that py3-htcondor spec has been downgraded to condor 8.9.7 from 9.2
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py3-htcondor.spec
https://github.com/cms-sw/cmsdist/commit/cfff97bd810ba33fe775a10d2e9eb66a69757604
while condorpy3.spec file has 9.2 binding
condorphttps://github.com/cms-sw/cmsdist/tree/comp_gcc630

can we use tokens with 8.9 API ?

**Comment by belforte** [comment]
@saqibhaleem fixed some authorizations in schedds and provided a test script[2] and helpted me to validate my token. I believe that we need full condor in TW, as in py2 versions. So rebuild TW with `condorpy3` in the spec file in place of `py3-condor` and it looks much better.
`condor_tokens_list` show the token and test script works from inside the new TW container in crab-dev-tw01

Now I get a ClassAdParseError exception[1] from unrelated lines where the `dagAd` is prepared, I guess condor 9.2 vs. condor 8.9, indeed the current code is quite strange, why use an ExprTree to set a simple integer value for CRAB_TaskSubmitTime ?
[1]
```
-> dagAd["CRAB_TaskSubmitTime"] = classad.ExprTree("%s" % info["start_time"].encode('ascii', 'ignore'))
(Pdb) 
classad.ClassAdParseError: <no detail available>
```
(also exception handling is bad here, no message is printed nor the exact locatio is reported, but this is a different story)
indeed the expression in () in the RHS evaluates to `classad.ExprTree("b'4543543'")` which has a very bad aspect, while the simple 
```
dagAd["CRAB_TaskSubmitTime"] = classad.ExprTree(info["start_time"])
```
gives no problem.

@mmascher what do you think ? Should we avoid ExprTree or simply avoid attempting to convert things with `encode` and similar since this is now all pure python3 and everything should better be unicode strrings ? @mapellidario feel free to add your thoughts

[2]
```
belforte@vocms055/HTCondor> cat remote_submit.py 
import os
import htcondor

collector = htcondor.Collector("vocms0803")
remote_schedd= collector.locate(htcondor.DaemonTypes.Schedd,"crab3@vocms068.cern.ch")
schedd = htcondor.Schedd(remote_schedd)
htcondor.param['SEC_CLIENT_AUTHENTICATION_METHODS'] = 'IDTOKENS,FS'

######### if need to use GSI authentication uncomment this
#htcondor.param['SEC_CLIENT_AUTHENTICATION_METHODS'] = 'GSI,FS'
#htcondor.param['DELEGATE_FULL_JOB_GSI_CREDENTIALS'] = 'true'
#htcondor.param['DELEGATE_JOB_GSI_CREDENTIALS_LIFETIME'] = '0'
#  Path to personal proxy
#os.environ['X509_USER_PROXY'] = '/tmp/myproxy'
##################
sub = htcondor.Submit('''
    Executable = ExampleOne.exe
    Hold = True
    +SubmitExample = True
    queue Arguments matching files *
    ''')
with schedd.transaction() as txn:
  rr=sub.queue(txn)
  print(rr)

```
Note: it would be better to make a new test script for our use with `schedd.submit` i.e. the same API used in the TW.

**Comment by belforte** [comment]
got a first taks running
https://cmsweb-test1.cern.ch/crabserver/ui/task/220115_155447%3Abelforte_crab_20220115_165115
from condor_q on vocms068:
```
826274.0   crabtw          1/15 16:57   0+00:01:00 R  0    0.0 task_proc_wrapper.sh 220115_155447
826275.0   crabtw          1/15 16:57   0+00:00:00 I  10   0.0 gWMS-CMSRunAnalysis.sh -a sandbox.
826276.0   crabtw          1/15 16:57   0+00:00:00 I  10   0.0 gWMS-CMSRunAnalysis.sh -a sandbox.
```

**Comment by belforte** [comment]
will close once https://github.com/dmwm/CRABServer/pull/6975 is merged in python3 branch

**Comment by belforte** [comment]
closed via #6975
